[BE] 트랜잭션 데이터 CSV/JSON 변환 로직 구현

KHJHK opened the pull request:
트랜잭션 로그들을 ElasticSearch에서 가져와 CSV/JSON 파일로 추출하는 로직 구현

## 변경점
1. DTO
- API 모듈에 CSV/JSON 추출 로직 구현을 위해 TransactionLog 위치를 simulator에서 core로 변경
- ES 저장 및 타입 지정을 위한 Document, Id, Field 어노테이션 추가
- ES에 저장 및 불러와지는 트랜잭션 데이터 타입을 맞추기 위해 timestamp 형식을 맞추기 위해 format 설정
2. TransactionController, TransactionExportService, ESTransactionSearchRepository
- 트랜잭션을 CSV/JSON으로 추출하는데 사용되는 controller, service, repository 구현
3. FileExporter util 기능
- 트랜잭션 로그 리스트를 CSV/JSON 형태로 변환하는 util class 구현
- 지수 형식으로 추출되는 balanced_before/after을 10진수로 추출하기 위해 ObjectMapperConfig에 지수표기->10진수 표기 직렬화 코드추가
4. TransactionService 테스트 코드 구현
5. logstash pipeline.conf의 불필요한 filter 구문 삭제

## 브라우저 테스트 내용
- 약 130만건 데이터 기준 CSV/JSON 파일 생성시 약 1분 내로 파일 생성 완료
- 파일 다운로드 도중 사용자가 정지시킬 경우 서버에서 에러 반환하며 클라이언트 연결 해제
- 클라이언트가 페이지를 이동해도, 브라우저의 다운로드 세션이 별개로 동작하여 다운로드 유지됨
 
## 체크리스트
- [x] 무엇을 변경했는지 설명
- [x] 테스트를 작성

## TODO
- 스트리밍 형식으로 데이터 전송하는 방식을 비동기 처리로 변경 or 서버 내부에서 데이터를 생성한 후 사용자에게 보내는 방식으로 변경

## 관련 이슈
- Closes #49

KHJHK committed:
#49 - chore : csv 출력을 위한 dependency 추가

csv 출력을 위한 dependency 추가

KHJHK committed:
#49 - refactor : TransactionLog timestamp 과련 dto 및 pipeline 수정

ES에 저장되는 TransactionLog 에서 저장이 배열 형태로 되던 timestamp를 직렬화 하여 저장
- DTO 및 logstash pipeline 변경

KHJHK committed:
#49 - chore : logstash pipeline 변경 및 불필요한 dependency 삭제

logstash pipeline 변경 및 불필요한 dependency 삭제

KHJHK committed:
Merge branch 'develop' of https://github.com/SIMPAYLOG/finance-data-generator into feat/49-transaction-to-csv

# Conflicts:
#	build.gradle
#	generator-simulator/src/main/java/com/simpaylog/generatorsimulator/dto/TransactionLog.java

KHJHK committed:
#49 - feat : ElasticSearch의 모든 데이터를 불러오기 구현

ElasticSearch의 모든 데이터를 userId순, 시간순으로 불러오는 기능 구현.
추후 csv||json 파일 생성에 사용

KHJHK committed:
#49 - test : TransactionExportServiceTest 생성 및 데이터 불러오기 기능 테스트

TransactionExportServiceTest 생성 및 데이터 불러오기 기능 테스트

KHJHK committed:
#49 - feat : TransactionLog to CSV/JSON 기능 구현 및 CSV/JSON 추출 로직 모듈 위치 변경

트랜잭션 로그를 csv/json 파일로 추출하는 기능 구현.

- FileExporter : TransactionLog들을 CSV/JSON 파일로 변환하는 util 객체

KHJHK committed:
#49 - refactor : TransactionLog CSV/JSON 추출 코드 리펙토링

TransactionLog CSV/JSON 추출 코드 리펙토링

- 지수 표기법으로 추출되던 balanced before/after 정보 10진수 표기법으로 직렬화
- CSV 파일의 각 컬럼명을 명시하기 위한 TransactionLogHeader.java Enum 추가
- TransactionExportServiceTest의 CSV/JSON 테스트에 모든 필드값이 들어갔는지 확인하는 로직 추가

KHJHK committed:
Merge branch 'develop' of https://github.com/SIMPAYLOG/finance-data-generator into feat/49-transaction-to-csv

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/service/TransactionExportService.java:
```diff
@@ -0,0 +1,62 @@
+            return new byte[0]; // 또는 예외 처리
+        }
+
+        try {
+            return switch (format.toLowerCase()) {
+                case "csv" -> fileExporter.toCSV(allTransactions);
+                case "json" -> fileExporter.toJSON(allTransactions);
+                default -> throw new IllegalArgumentException("파일 변환을 지원하지 않는 형식입니다.: " + format);
+            };
```
컨트롤러에서랑 중복되는 검증로직인 것 같은데 맞나요?
커스텀 exception 사용해주세요.

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/service/TransactionExportService.java:
```diff
@@ -0,0 +1,62 @@
+                case "json" -> fileExporter.toJSON(allTransactions);
+                default -> throw new IllegalArgumentException("파일 변환을 지원하지 않는 형식입니다.: " + format);
+            };
+        } catch (IOException e) {
+            throw new RuntimeException("데이터 변환중 에러가 발생했습니다.", e);
```
커스텀 익셉션, server error(500) 으로 남겨주세요.

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/service/TransactionExportService.java:
```diff
@@ -0,0 +1,62 @@
+                page++;
+            }
+        } catch (Exception e){
+            System.out.println("TransactionExportService 에러 발생 : " + e.getMessage());
```
slf4j 사용 바랍니다.

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/service/TransactionExportService.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    public List<TransactionLog> getAllTransaction() {
```
데이터가 너무 많다면 비동기 처리로 진행하고, 결과값 모아서 합치는 작업으로 할당해도 좋을 것 같아요. 다만 커넥션 풀을 신경써서 할당해줘야할듯 합니다.

lmw7414 commented on generator-api/src/test/java/com/simpaylog/generatorapi/service/TransactionExportServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+    @MockitoBean
+    RedisRepository redisRepository;
+
+    @Test
+    public void ES에서_모든_데이터를_가져온다() {
+        List<TransactionLog> results = transactionExportService.getAllTransaction();
+
+        assertThat(results)
+                .withFailMessage("Elasticsearch에서 조회된 결과가 null입니다.")
+                .isNotNull();
+
+        assertThat(results)
+                .withFailMessage("Elasticsearch에서 결과를 가져왔지만 비어 있습니다.")
+                .isNotEmpty();
+
+        TransactionLog first = results.getFirst();
+        System.out.println(first);
+
+        assertThat(first.userId())
+                .withFailMessage("첫 번째 데이터의 userId가 null입니다.")
+                .isNotNull();
+
+        assertThat(first.timestamp())
+                .withFailMessage("첫 번째 데이터의 timestamp가 null입니다.")
+                .isNotNull();
+
+        System.out.println("가져온 데이터 수: " + results.size());
+        System.out.println("첫 번째 데이터 userId: " + first.userId());
+        System.out.println("첫 번째 데이터 timestamp: " + first.timestamp());
+    }
```
테스트의 의미가 모호한 것 같습니다. 
반환할 데이터의 mock 객체를 미리 만들어 놓고 테스트하여 검증하는 방식으로 하는 편이 좋아보여요.
현재는 ES에 어떠한 결과를 예상하고 테스트하려고 하는지 그림이 그려지지 않습니다. 뿐만 아니라 테스트 메서드의 이름과 검증하고자 하는 결과가 무엇인지 모르겠어요. 
1. 정상적인 케이스
2. 데이터를 조회했을 때 없을 경우의 테스트 케이스
3. 예측했던 데이터의 필드 값이 없을 때의 테스트 케이스
이렇게 세부 항목으로 여러번 쪼개서 하는 편이 나을 것 같습니다.

lmw7414 commented on generator-api/src/test/java/com/simpaylog/generatorapi/service/TransactionExportServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+                    }
+                });
+
+//    System.out.println("JSON 결과: " + json.substring(0, Math.min(500, json.length())) + "...");
```
불필요한 print문은 지워도 좋을 것 같아요.

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/dto/TransactionLog.java:
ES 용 Kafka 용 DTO를 서로 나눠 관리하는 편이 좋아보입니다.

lmw7414 commented on generator-core/src/main/java/com/simpaylog/generatorcore/enums/TransactionLogHeader.java:
해당 클래스 명만 봤을 때구체적인 활용 방법에 대해 잘 모르겠습니다.
클래스명을 구체적으로 해도 되고, 별도의 패키지로 분리해서 명확히 해주는 편이 좋을 것 같아요.

KHJHK committed:
#49 - feat : CSV/JSON 대용량 데이터 추출 로직 구현

기존 1만건 이상 데이터를 JSON/CSV로 추출시 데이터가 누락됨을 확인하여, 대용량 데이터를 JSON/CSV로 추출 가능하도록 로직 구현

- TransactionLog를 내부 dto로 사용하고자, Elasticsearch와 통신하는 dto인 TransactionLogDocument 구현
- 스트리밍 + 콜백 함수(Consumer)를 통한 대용량 JSON/CSV 다운로드 기능 구현
- request-timeout을 길게 설정하여 대용량 데이터 전송시 스트리밍 연결 가능하도록 설정
- 기존 ES에서 제공하던 간편화된 쿼리 사용시 1만건 이상 데이터 조회가 불가능하여, custom query가 가능하도록 rpository 변경

KHJHK committed:
#49 - refactor : CSV/JSON 파일 추출 코드 리펙토링

CSV/JSON 파일 추출 코드 리펙토링
- 파일 추출 타입(CSV, JSON)을 명시하는 ExportFormat Enum 추가
- 커스텀 에러 처리를 사용하여 에러 처리
- 중복되는 에러 처리 코드 삭제
- TransactionLogHeader의 이름을 TransactionCsvExportHeader로 변경하여 더 명시적으로 변경

KHJHK committed:
Merge branch 'develop' of https://github.com/SIMPAYLOG/finance-data-generator into feat/49-transaction-to-csv

KHJHK committed:
#49 - feat : sessionId를 기반으로 CSV/JSON 파일 추출

- sessionId를 기반으로 CSV/JSON 파일을 추출하도록 변경
- es 쿼리로 데이터 가져올시, 중복 시간대 발생한 트랜잭션으로 인해 데이터가 누락됨을 방지하기 위해 uuid 기준 정렬점 추가

KHJHK committed:
#49 - test : TransactionExportServiceTest코드 작성

- TransactionExportService 테스트 코드 작성

KHJHK committed:
#49 - chore : TransactionExportServiceTest 불필요한 properties 삭제

TransactionExportServiceTest 불필요한 properties 삭제

lmw7414 commented on generator-api/src/main/java/com/simpaylog/generatorapi/controller/TransactionController.java:
```diff
@@ -0,0 +1,49 @@
+public class TransactionController {
+    private final TransactionExportService transactionExportService;
+
+    @GetMapping
```
URL이 조금 더 명확할 필요가 있어보입니다.

lmw7414 commented on generator-api/src/main/resources/application-api.yml:
```diff
@@ -3,4 +3,7 @@ server:
+    name: generator-api
+  mvc:
+    async:
+      request-timeout: 500000
```
무서워요...

KHJHK replied:
무섭긴해요...

lmw7414 reviewed (approved):
클라이언트의 예측할 수 없는 예외 상황에 대한 대비가 어려워 보입니다. 사용자 요청이 몰릴 시 서버 부하가 올 것 같아요.

KHJHK committed:
#49 - chore : logstash 설정 변경 및 dto 변경

es에 저장되는 데이터 타입 변경을 위해 logstash 파이프라인 수정 및 dto 수정

KHJHK committed:
#49 - chore : uuid keyword 기반 정렬 누락부분 수정

text 타입인 uuid keyword 기반 정렬이 누락되어 에러나던 코드 수정

KHJHK committed:
#49 - chore : controller url 수정 및 데이터 형식 검증 위치 변경

- api/transactions -> api/transactions/export로 수정

- format이 css/json인지 검증하는 로직을 service로 이동 및 에러 코드 추가

KHJHK committed:
#49 - chore : logstash 설정 파일 원복

- 불명확한 logstash 설정 파일들 삭제 및 logstash docker-compose 설정 원상복구

KHJHK committed:
#49 - refactor : ElasticsearchRepository uuid, sessionId 관련 쿼리 변경

uuid, sessionId의 매핑 정의를 text -> keyword로 변경하여 관련 쿼리 변경

KHJHK committed:
#49 - refactor : Transaction timestamp 표기 시간초 추가

트랜잭션 dto들의 timestamp 표기를 yyyy-MM-dd HH:mm -> yyyy-MM-dd HH:mm:ss 변경

KHJHK committed:
#49 - chose : logstash filter 시간초 누락 변경

logstash filter 시간초 누락 변경
